pkg/workflows/management/core_components.go

```diff
@@ -0,0 +1,110 @@
+package management
+
+import (
+	"context"
+
+	"github.com/aws/eks-anywhere/pkg/logger"
+	"github.com/aws/eks-anywhere/pkg/task"
+	"github.com/aws/eks-anywhere/pkg/types"
+	"github.com/aws/eks-anywhere/pkg/workflows"
+)
+
+type ensureEtcdCAPIComponentsExist struct{}
+
+// Run ensureEtcdCAPIComponentsExist ensures ETCD CAPI providers on the management cluster.
+func (s *ensureEtcdCAPIComponentsExist) Run(ctx context.Context, commandContext *task.CommandContext) task.Task {
+	logger.Info("Ensuring etcd CAPI providers exist on management cluster before upgrade")
+	if err := commandContext.CAPIManager.EnsureEtcdProvidersInstallation(ctx, commandContext.ManagementCluster, commandContext.Provider, commandContext.CurrentClusterSpec); err != nil {
+		commandContext.SetError(err)
+		return &workflows.CollectMgmtClusterDiagnosticsTask{}
+	}
+	return &pauseGitOpsReconcile{}
+}
+
+func (s *ensureEtcdCAPIComponentsExist) Name() string {
+	return "ensure-etcd-capi-components-exist"
+}
+
+func (s *ensureEtcdCAPIComponentsExist) Checkpoint() *task.CompletedTask {
+	return &task.CompletedTask{
+		Checkpoint: nil,
+	}
+}
+
+func (s *ensureEtcdCAPIComponentsExist) Restore(ctx context.Context, commandContext *task.CommandContext, completedTask *task.CompletedTask) (task.Task, error) {
+	return &pauseGitOpsReconcile{}, nil
+}
+
+type upgradeCoreComponents struct {
+	UpgradeChangeDiff *types.ChangeDiff
+}
+
+// Run upgradeCoreComponents upgrades pre cluster upgrade components.
+func (s *upgradeCoreComponents) Run(ctx context.Context, commandContext *task.CommandContext) task.Task {
+	logger.Info("Upgrading core components")
+
+	err := commandContext.Provider.PreCoreComponentsUpgrade(
+		ctx,
+		commandContext.ManagementCluster,
+		commandContext.ClusterSpec,
+	)
+	if err != nil {
+		commandContext.SetError(err)
+		return &workflows.CollectMgmtClusterDiagnosticsTask{}
+	}
+
+	changeDiff, err := commandContext.CAPIManager.Upgrade(ctx, commandContext.ManagementCluster, commandContext.Provider, commandContext.CurrentClusterSpec, commandContext.ClusterSpec)
+	if err != nil {
+		commandContext.SetError(err)
+		return &workflows.CollectMgmtClusterDiagnosticsTask{}
+	}
+	commandContext.UpgradeChangeDiff.Append(changeDiff)
+
+	if err = commandContext.GitOpsManager.Install(ctx, commandContext.ManagementCluster, commandContext.CurrentClusterSpec, commandContext.ClusterSpec); err != nil {
+		commandContext.SetError(err)
+		return &workflows.CollectMgmtClusterDiagnosticsTask{}
+	}
+
+	changeDiff, err = commandContext.GitOpsManager.Upgrade(ctx, commandContext.ManagementCluster, commandContext.CurrentClusterSpec, commandContext.ClusterSpec)
+	if err != nil {
+		commandContext.SetError(err)
+		return &workflows.CollectMgmtClusterDiagnosticsTask{}
+	}
```

Comment:
This looks a little weird? Why is GitOps being installed and then upgraded right away?
If it's being installed fresh, shouldn't the latest version be installed already?

Reply:
I guess we install flux controllers during core components upgrade if it does not exist before in the cluster spec.  https://github.com/aws/eks-anywhere/pull/2759
It is a part of current workflow https://github.com/aws/eks-anywhere/blob/main/pkg/workflows/upgrade.go#L251-L260 and we haven't changed it.